scalameter-core/src/main/scala/org/scalameter/InvocationCountMatcher.scala

```diff
@@ -0,0 +1,123 @@
+package org.scalameter
+
+import java.lang.reflect.Method
+import java.util.regex.Pattern
+import org.objectweb.asm.Type
+import scala.util.matching.Regex
+
+
+sealed trait ClassMatcher {
+  /** Matches class.
+   *
+   *  Note that any '/' in a class name is replaced by a '.'.
+   */
+  def matches(className: String): Boolean
+}
+
+/** Matches class with a class name given as a string. */
+case class StringClassMatcher(clazz: String) extends ClassMatcher {
+  def this(clazz: Class[_]) = this(clazz.getName)
+  
+  def matches(className: String): Boolean = className.replace('/', '.') == clazz
+}
+
+object StringClassMatcher {
+  def apply(clazz: Class[_]) = new StringClassMatcher(clazz)
+}
+
+/** Matches class with a regex. */
+case class RegexClassMatcher(regex: Pattern) extends ClassMatcher {
+  def matches(className: String): Boolean = regex.matcher(className.replace('/', '.')).matches()
+}
+
+sealed trait MethodMatcher {
+  /** Matches method.
+   *
+   *  @param methodName method name
+   *  @param methodDescriptor method descriptor
+   *
+   *  @see https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.3
+   */
+  def matches(methodName: String, methodDescriptor: String): Boolean
+}
+
+/** Matches method with a method name given as a string. */
+case class StringMethodMatcher(method: String) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean = methodName == method
+}
+
+/** Matches method with a regex. */
+case class RegexMethodMatcher(regex: Pattern) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean = regex.matcher(methodName).matches()
+}
+
+/** Matches method with a [[java.lang.reflect.Method]].
+ *
+ *  That means that method name, method arguments and method return type are matched.
+ */
+case class FullMethodMatcher(method: Method) extends MethodMatcher {
+  def matches(methodName: String, methodDescriptor: String): Boolean =
+    methodName == method.getName && methodDescriptor == Type.getMethodDescriptor(method)
+}
+
+
+/** Pattern used for matching methods whose invocations should be counted.
+ *
+ *  @param classMatcher matches full class name given in internal format ('/' instead of a '.' as a separator)
+ *  @param methodMatcher matches method name
+ */
+case class InvocationCountMatcher(classMatcher: ClassMatcher, methodMatcher: MethodMatcher) {
+  /** Matches allocations of a class. */
+  def this(clazz: Class[_]) = this(StringClassMatcher(clazz), StringMethodMatcher("<init>"))
```

Comment:
The Java API can always pay the price for their choice of language by boilerplating with the most generic primary ctor :)

I wouldn't optimize for the `japi`, since it is already verbose.